model/src/application/interaction/application_command/data/mod.rs

```diff
@@ -26,50 +30,167 @@ pub struct CommandData {
+struct CommandDataOptionRaw<'a> {
+    name: String,
+    #[serde(rename = "type")]
+    kind: CommandOptionType,
+    value: Option<CommandOptionValueRaw<'a>>,
+    #[serde(default)]
+    options: Option<Vec<CommandDataOption>>,
+}
+
+#[derive(Debug, Deserialize, Serialize)]
+enum CommandOptionValueRaw<'a> {
+    String(Cow<'a, str>),
+    Integer(i64),
+    Boolean(bool),
+}
+
+impl Serialize for CommandDataOption {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut state = serializer.serialize_struct("CommandDataOptionRaw", 3)?;
+        state.serialize_field("name", &self.name)?;
+        state.serialize_field("type", &self.value.kind())?;
+        match self.value {
+            CommandOptionValue::SubCommand(ref opts)
+            | CommandOptionValue::SubCommandGroup(ref opts) => {
+                state.serialize_field("options", &Some(opts))?
+            }
+            CommandOptionValue::String(ref s) => {
+                state.serialize_field("value", &Some(CommandOptionValueRaw::String(s.into())))?
+            }
+            CommandOptionValue::Integer(i) => {
+                state.serialize_field("value", &Some(CommandOptionValueRaw::Integer(i)))?
+            }
+            CommandOptionValue::Boolean(b) => {
+                state.serialize_field("value", &Some(CommandOptionValueRaw::Boolean(b)))?
+            }
+            CommandOptionValue::User(UserId(id))
+            | CommandOptionValue::Channel(ChannelId(id))
+            | CommandOptionValue::Role(RoleId(id))
+            | CommandOptionValue::Mentionable(GenericId(id)) => state.serialize_field(
+                "value",
+                &Some(CommandOptionValueRaw::String(id.to_string().into())),
+            )?,
         }
+        state.end()
     }
+}
+
+impl<'de> Deserialize<'de> for CommandDataOption {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let raw = CommandDataOptionRaw::deserialize(deserializer)?;
+        let value = if let Some(value) = raw.value {
+            match (raw.kind, value) {
+                (CommandOptionType::String, CommandOptionValueRaw::String(s)) => {
+                    CommandOptionValue::String(s.into_owned())
+                }
+                (CommandOptionType::Integer, CommandOptionValueRaw::Integer(i)) => {
+                    CommandOptionValue::Integer(i)
+                }
+                (CommandOptionType::Boolean, CommandOptionValueRaw::Boolean(b)) => {
+                    CommandOptionValue::Boolean(b)
+                }
+                (CommandOptionType::User, CommandOptionValueRaw::String(s)) => {
+                    let id = UserId(s.parse().map_err(|_| {
+                        de::Error::invalid_value(de::Unexpected::Str(&s), &"user ID")
+                    })?);
+                    CommandOptionValue::User(id)
```

Comment:
return statements should have a newline above them, please check this entire match statement

Reply:
Sorry, I'm not sure I understood this correctly. Does 8dca42b address your comments?

Reply:
yes, thank you